programs/coherence-beamsplitter/src/state.rs

```diff
@@ -1,44 +1,69 @@
+use std::mem::size_of;
+
 use anchor_lang::prelude::*;
 
-/// Contains the info of the prism etf. Immutable.
+#[constant]
+pub const _PRISM_ETF_SIZE: usize = size_of::<PrismEtf>();
+#[constant]
+pub const MAX_WEIGHTED_TOKENS: usize = 10;
+
+/// Contains the info of the prism etf.
 #[account]
-#[derive(Debug)]
+#[derive(Copy, Debug, Default)]
 pub struct PrismEtf {
-    /// The Beamsplitter metadata account associated with this prism etf
-    pub prism_etf: Pubkey,
-    /// Authority that has admin rights over the [PrismEtf].
-    pub authority: Pubkey,
-    /// Bump seed. Stored for find_program_address on-chain performance
+    /// Manager that has reblancing / closing rights over the [PrismEtf].
+    pub manager: Pubkey,
+    /// The account containing the [WeightedTokens] of this [PrismEtf]
+    pub weighted_tokens: Pubkey,
+    /// If true, the ETF can no longer have assets added to it (without rebalancing)
+    pub is_finished: bool,
+    /// The bump for this PDA account
     pub bump: u8,
-    /// [Mint] of the [PrismEtf]
-    pub mint: Pubkey,
-    // TODO: replace 8 with shared max size
-    // TODO: find way to have optional serialization
-    /// [AssetSource] array
-    pub weighted_tokens: [WeightedToken; 8],
 }
 
-#[derive(AnchorDeserialize, AnchorSerialize, Default, Debug, Clone, Copy)]
-pub struct WeightedToken {
-    pub mint: Pubkey,
-    pub weight: u64,
+#[account(zero_copy)]
+#[derive(Debug)]
+pub struct WeightedTokens {
+    /// The index of array
+    pub index: u32,
+    /// Max capacity of the array
+    pub capacity: u32,
+    /// [WeightedToken] array
+    pub weighted_tokens: [WeightedToken; 10], // TODO find better name
 }
 
-impl Default for PrismEtf {
-    #[inline(never)]
-    fn default() -> PrismEtf {
-        // TODO replace 8 with shared max size
-        PrismEtf {
-            prism_etf: Pubkey::default(),
-            authority: Pubkey::default(),
-            bump: u8::default(),
-            mint: Pubkey::default(),
-            weighted_tokens: [WeightedToken {
-                mint: Pubkey::default(),
-                weight: 0,
-            }; 8],
-        }
-    }
+/// Contains the info of the prism etf.
+#[account]
+#[derive(Copy, Debug, Default)]
+pub struct OrderState {
+    /// [WeightedToken] array
+    pub transferred_tokens: Pubkey,
+    /// If true, the order is CONSTRUCTION type, otherwise DECONSTRUCTION
+    pub is_construction: bool, // TODO use enum
```

Comment:
Might want to setup this enum before we merge

Reply:
Good point